### PR TITLE
[Enhance] Add type ints in these files:

### PR DESCRIPTION
**Pull request by WINDSKY45**
## Modification

Add type hints in these files:
'base_module.py', 'checkpoint.py',
'default_constructor.py',
'dist_utils.py', 'fp16_utils.py',
'log_buffer.py', 'priority.py'.

## Checklist

**Before PR**:

- [x] I have read and followed the workflow indicated in the [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) to create this PR.
- [x] Pre-commit or linting tools indicated in [CONTRIBUTING.md](https://github.com/open-mmlab/mmcv/blob/master/CONTRIBUTING.md) are used to fix the potential lint issues.
- [x] Bug fixes are covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] New functionalities are covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, including docstring or example tutorials.

**After PR**:

- [x] If the modification has potential influence on downstream or other related projects, this PR should be tested with some of those projects, like MMDet or MMCls.
- [x] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmcv/runner/checkpoint.py [hunk 1]
```diff
@@ -220,7 +227,10 @@ def _register_scheme(cls, prefixes, loader, force=False):
             sorted(cls._schemes.items(), key=lambda t: t[0], reverse=True))
 
     @classmethod
-    def register_scheme(cls, prefixes, loader=None, force=False):
+    def register_scheme(cls,
+                        prefixes: Union[str, List[str], Tuple[str, ...]],
```
Comment (HAOCHENYE) [inline mmcv/runner/checkpoint.py:231]: Suggest to use `Sequence[str]` to replace `List[str], Tuple[str, ...]`, and modify the docstring with prefixes (str, Sequence[str])

File: mmcv/runner/checkpoint.py [hunk 2]
```diff
@@ -363,7 +385,10 @@ def load_from_pavi(filename, map_location=None):
 
 
 @CheckpointLoader.register_scheme(prefixes=r'(\S+\:)?s3://')
-def load_from_ceph(filename, map_location=None, backend='petrel'):
+def load_from_ceph(
+        filename: str,
+        map_location: Optional[str] = None,
+        backend: Optional[str] = 'petrel') -> Union[dict, OrderedDict]:
```
Comment (HAOCHENYE) [inline mmcv/runner/checkpoint.py:391]: Optional should only be used when default value is `None`

File: mmcv/runner/dist_utils.py [hunk 3]
```diff
@@ -136,12 +138,14 @@ def wrapper(*args, **kwargs):
...
-def allreduce_params(params, coalesce=True, bucket_size_mb=-1):
+def allreduce_params(params: List[Parameter],
+                     coalesce: bool = True,
+                     bucket_size_mb: int = -1):
     """Allreduce parameters.
 
     Args:
-        params (list[torch.Parameters]): List of parameters or buffers of a
-            model.
+        params (list[torch.nn.parameter.Parameter]): List of parameters or
```
Comment (HAOCHENYE) [inline mmcv/runner/dist_utils.py:147]:
```suggestion
        params (List[torch.nn.parameter.Parameter]): List of parameters or
```

File: mmcv/runner/fp16_utils.py [hunk 4]
```diff
@@ -21,7 +23,7 @@
     pass
 
 
-def cast_tensor_type(inputs, src_type, dst_type):
+def cast_tensor_type(inputs, src_type: torch.dtype, dst_type: torch.dtype):
```
Comment (HAOCHENYE) [inline mmcv/runner/fp16_utils.py:26]: We can set return type with any to avoid use `type: ignore`

File: mmcv/runner/fp16_utils.py [hunk 5]
```diff
@@ -52,18 +54,20 @@ def cast_tensor_type(inputs, src_type, dst_type):
...
     elif isinstance(inputs, abc.Iterable):
-        return type(inputs)(
+        return type(inputs)(  # type: ignore
             cast_tensor_type(item, src_type, dst_type) for item in inputs)
     else:
         return inputs
 
 
-def auto_fp16(apply_to=None, out_fp32=False, supported_types=(nn.Module, )):
+def auto_fp16(apply_to: Optional[Iterable] = None,
```
Comment (HAOCHENYE) [inline mmcv/runner/fp16_utils.py:68]: The decorator needs a return value, Maybe `Callable`?

File: mmcv/runner/fp16_utils.py [hunk 6]
```diff
@@ -150,7 +154,7 @@ def new_func(*args, **kwargs):
     return auto_fp16_wrapper
 
 
-def force_fp32(apply_to=None, out_fp16=False):
+def force_fp32(apply_to: Optional[Iterable] = None, out_fp16: bool = False):
```
Comment (HAOCHENYE) [inline mmcv/runner/fp16_utils.py:157]: same as above

**Review by HAOCHENYE — COMMENTED** [review]
We don't need to add return Type Hints for `__init__` method.

File: mmcv/runner/fp16_utils.py (hunk 4 above)
Comment (WINDSKY45) [reply mmcv/runner/fp16_utils.py:26]: I tried it but can't fix it.

**Comment by WINDSKY45** [comment]
I have fixed them.

File: mmcv/runner/checkpoint.py [hunk 7]
```diff
@@ -220,13 +227,16 @@ def _register_scheme(cls, prefixes, loader, force=False):
...
+                        prefixes: Union[str, Sequence[str]],
+                        loader: Optional[Callable] = None,
+                        force: Optional[bool] = False):
         """Register a loader to CheckpointLoader.
 
         This method can be used as a normal class method or a decorator.
 
         Args:
-            prefixes (str or list[str] or tuple[str]):
+            prefixes (str or Sequence[str]):
```
Comment (zhouzaida) [inline mmcv/runner/checkpoint.py:239]: A string is also a sequence so should we use Sequnece here?

File: mmcv/runner/dist_utils.py [hunk 8]
```diff
@@ -158,11 +162,14 @@ def allreduce_params(params, coalesce=True, bucket_size_mb=-1):
             dist.all_reduce(tensor.div_(world_size))
 
 
-def allreduce_grads(params, coalesce=True, bucket_size_mb=-1):
+def allreduce_grads(params: List[torch.nn.parameter.Parameter],
```
Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:165]:
```suggestion
def allreduce_grads(params: List[torch.nn.Parameter],
```

File: mmcv/runner/dist_utils.py [hunk 9]
```diff
@@ -158,11 +162,14 @@ def allreduce_params(params, coalesce=True, bucket_size_mb=-1):
...
+                    coalesce: bool = True,
+                    bucket_size_mb: int = -1):
     """Allreduce gradients.
 
     Args:
-        params (list[torch.Parameters]): List of parameters of a model
+        params (list[torch.nn.parameter.Parameters]): List of parameters of
```
Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:171]:
```suggestion
        params (list[torch.nn.Parameters]): List of parameters of
```

File: mmcv/runner/fp16_utils.py [hunk 10]
```diff
@@ -52,18 +54,22 @@ def cast_tensor_type(inputs, src_type, dst_type):
...
     else:
         return inputs
 
 
-def auto_fp16(apply_to=None, out_fp32=False, supported_types=(nn.Module, )):
+def auto_fp16(
+        apply_to: Optional[Iterable] = None,
+        out_fp32: bool = False,
+        supported_types: tuple = (nn.Module, ),
+) -> Callable:
```
Comment (zhouzaida) [inline mmcv/runner/fp16_utils.py:72]: Can we use the following format:
```python
def auto_fp16(apply_to: Optional[Iterable] = None,
              out_fp32: bool = False,
              supported_types: tuple = (nn.Module, )) -> Callable:
```

File: mmcv/runner/fp16_utils.py [hunk 11]
```diff
@@ -236,15 +243,17 @@ def new_func(*args, **kwargs):
     return force_fp32_wrapper
 
 
-def allreduce_grads(params, coalesce=True, bucket_size_mb=-1):
-    warnings.warning(
+def allreduce_grads(params: List[Parameter],
+                    coalesce: bool = True,
+                    bucket_size_mb: int = -1):
```
Comment (zhouzaida) [inline mmcv/runner/fp16_utils.py:248]:
```suggestion
                    bucket_size_mb: int = -1) -> None:
```

File: mmcv/runner/fp16_utils.py [hunk 12]
```diff
@@ -359,7 +371,7 @@ def __init__(self,
         self.scale_factor = scale_factor
         self.scale_window = scale_window
 
-    def has_overflow(self, params):
+    def has_overflow(self, params: List[Parameter]):
```
Comment (zhouzaida) [inline mmcv/runner/fp16_utils.py:374]:
```suggestion
    def has_overflow(self, params: List[Parameter]) -> bool:
```

File: mmcv/runner/fp16_utils.py [hunk 13]
```diff
@@ -382,7 +394,7 @@ def _has_inf_or_nan(x):
                 return True
             return False
 
-    def update_scale(self, overflow):
+    def update_scale(self, overflow: bool):
```
Comment (zhouzaida) [inline mmcv/runner/fp16_utils.py:397]:
```suggestion
    def update_scale(self, overflow: bool) -> None:
```

File: mmcv/runner/fp16_utils.py [hunk 14]
```diff
@@ -405,7 +417,7 @@ def state_dict(self):
             scale_factor=self.scale_factor,
             scale_window=self.scale_window)
 
-    def load_state_dict(self, state_dict):
+    def load_state_dict(self, state_dict: dict):
```
Comment (zhouzaida) [inline mmcv/runner/fp16_utils.py:420]:
```suggestion
    def load_state_dict(self, state_dict: dict) -> None:
```

File: mmcv/runner/log_buffer.py [hunk 15]
```diff
@@ -21,7 +21,7 @@ def clear_output(self):
         self.output.clear()
         self.ready = False
 
-    def update(self, vars, count=1):
+    def update(self, vars: dict, count: int = 1):
```
Comment (zhouzaida) [inline mmcv/runner/log_buffer.py:24]:
```suggestion
    def update(self, vars: dict, count: int = 1) -> None:
```

File: mmcv/runner/log_buffer.py [hunk 16]
```diff
@@ -30,7 +30,7 @@ def update(self, vars, count=1):
             self.val_history[key].append(var)
             self.n_history[key].append(count)
 
-    def average(self, n=0):
+    def average(self, n: int = 0):
```
Comment (zhouzaida) [inline mmcv/runner/log_buffer.py:33]:
```suggestion
    def average(self, n: int = 0) -> None:
```

File: mmcv/runner/dist_utils.py [hunk 17]
```diff
@@ -33,7 +35,7 @@ def _is_free_port(port):
         return all(s.connect_ex((ip, port)) != 0 for ip in ips)
 
 
-def init_dist(launcher, backend='nccl', **kwargs):
+def init_dist(launcher: str, backend: str = 'nccl', **kwargs):
```
Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:38]:
```suggestion
def init_dist(launcher: str, backend: str = 'nccl', **kwargs) -> None:
```

File: mmcv/runner/dist_utils.py [hunk 18]
```diff
@@ -125,7 +127,7 @@ def get_dist_info():
     return rank, world_size
 
 
-def master_only(func):
+def master_only(func: Callable):
```
Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:130]:
```suggestion
def master_only(func: Callable) -> Callable:
```

File: mmcv/runner/dist_utils.py [hunk 19]
```diff
@@ -136,12 +138,14 @@ def wrapper(*args, **kwargs):
     return wrapper
 
 
-def allreduce_params(params, coalesce=True, bucket_size_mb=-1):
+def allreduce_params(params: List[torch.nn.Parameter],
+                     coalesce: bool = True,
+                     bucket_size_mb: int = -1):
```
Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:143]:
```suggestion
                     bucket_size_mb: int = -1) -> None:
```

File: mmcv/runner/dist_utils.py [hunk 20]
```diff
@@ -158,11 +162,13 @@ def allreduce_params(params, coalesce=True, bucket_size_mb=-1):
             dist.all_reduce(tensor.div_(world_size))
 
 
-def allreduce_grads(params, coalesce=True, bucket_size_mb=-1):
+def allreduce_grads(params: List[torch.nn.Parameter],
+                    coalesce: bool = True,
+                    bucket_size_mb: int = -1):
```
Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:167]:
```suggestion
                    bucket_size_mb: int = -1) -> None:
```

**Comment by zhouzaida** [comment]
Please add type hints for `get_dist_info` in `mmcv/runner/dist_utils.py`.

**Comment by zhouzaida** [comment]
Hi, please resolve conflicts.

File: mmcv/runner/hooks/logger/text.py [hunk 21]
```diff
@@ -200,7 +204,7 @@ def _round_float(self, items):
         else:
             return items
 
-    def log(self, runner):
+    def log(self, runner: BaseRunner) -> dict:
```
Comment (zhouzaida) [inline mmcv/runner/hooks/logger/text.py:207]: We do not need to add a type hint for the runner.

**Review by zhouzaida — APPROVED** [review]

**Review by teamwong111 — APPROVED** [review]

File: mmcv/runner/dist_utils.py [hunk 22]
```diff
@@ -5,9 +5,11 @@
 import socket
 import subprocess
 from collections import OrderedDict
+from typing import Callable, List, Optional, Tuple
 
 import torch
 import torch.multiprocessing as mp
+import torch.nn
```
Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:12]: There is no need to import `torch.nn`.

Comment (zhouzaida) [inline mmcv/runner/dist_utils.py:12]: There is no need to import `torch.nn`.

File: mmcv/runner/checkpoint.py [hunk 23]
```diff
@@ -9,8 +10,10 @@
 from collections import OrderedDict
 from importlib import import_module
 from tempfile import TemporaryDirectory
+from typing import Callable, List, Optional, Sequence, Union
 
 import torch
+import torch.nn
```
Comment (zhouzaida) [inline mmcv/runner/checkpoint.py:16]: There is no need to import `torch.nn`.

Comment (zhouzaida) [inline mmcv/runner/checkpoint.py:16]: There is no need to import `torch.nn`.

File: mmcv/runner/checkpoint.py [hunk 24]
```diff
@@ -220,13 +226,16 @@ def _register_scheme(cls, prefixes, loader, force=False):
             sorted(cls._schemes.items(), key=lambda t: t[0], reverse=True))
 
     @classmethod
-    def register_scheme(cls, prefixes, loader=None, force=False):
+    def register_scheme(cls,
+                        prefixes: Union[str, Sequence[str]],
+                        loader: Optional[Callable] = None,
+                        force: Optional[bool] = False):
```
Comment (zhouzaida) [inline mmcv/runner/checkpoint.py:232]:
```suggestion
                        force: bool = False):
```

**Review by zhouzaida — APPROVED** [review]

**Pull request merged**